internal/dag/builder_test.go

```diff
@@ -8452,7 +8471,14 @@ func TestDAGInsert(t *testing.T) {
 			objs: []interface{}{
 				proxy103, proxy103a, s1,
 			},
-			want: listeners(),
+			want: listeners(
```

Comment:
This should still result in an error right? You can't specify multiple prefixes on a route. Additionally, the one's specified in the proxyies are for `/v1` and `/api`, so did `/` match?

Reply:
True, something fishy here, sorry for missing this! Will come back shortly, after going through this with a debugger.

Reply:
Didn't need debugger but it seems some of the test cases get bit tricky. Even if I just gone through this one I still got confused a few days after :)  In this case we have two proxies:  root proxy `proxy103` and child proxy `proxy103a`. Note that the [root proxy](https://github.com/projectcontour/contour/blob/0a94c00f2bf748e1afee914c8ce5fe06b0707a60/internal/dag/builder_test.go#L5674) has both `includes` **and also** `routes` section:

* Include section is invalid because there is two prefix conditions.  Therefore child `proxy103a` is not considered at all.
* Route section to `s1` is valid. It has no prefix condition i.e. prefix is implied `/`

Since the [route section](https://github.com/projectcontour/contour/blob/0a94c00f2bf748e1afee914c8ce5fe06b0707a60/internal/dag/builder_test.go#L5692) is valid we got listener for routing everything to `s1`. 

So I think from this perspective it was correct after all.  Maybe the test could be made bit easier to understand, for example: 

* Maybe errors in child proxy could be removed. They serve no meaning for the result of this test but may confuse reader.
* Don't use exactly same `/v1` and `/api` prefixes in both [root proxy](https://github.com/projectcontour/contour/blob/0a94c00f2bf748e1afee914c8ce5fe06b0707a60/internal/dag/builder_test.go#L5674) and the [child proxy](https://github.com/projectcontour/contour/blob/0a94c00f2bf748e1afee914c8ce5fe06b0707a60/internal/dag/builder_test.go#L5701).  If we did not have error due to multiple includes, full prefix could be doubled, like `/v1/v1` which might unnecessarily confuse the reader.

Reply:
If I change the test case by removing the double prefix conditions and I leave `/api` to root proxy, `/v1` to child proxy, then I get listener with

* `/` routed to `s1`
* `/api/v1` with 503 direct response

reader maybe would have expected route `/api/v1` to be valid and direct to `s1` as well, but now the problem is that child proxy is create in namespace `teama`, while services `s1` is in `default`. So maybe two more suggestions to make the test easier to understand:

* Have two different services, so route in root and route in child proxy do not refer to same service
* Create the second service in `teama` namespace and have child refer to that

Reply:
I've now updated existing test case "insert httpproxy with multiple prefix conditions on include" just for bit more clarity.